crates/bevy_reflect/src/type_registry.rs

```diff
@@ -182,27 +182,30 @@ impl TypeRegistration {
     /// data.
     ///
     /// Returns `None` if no such value exists.
+    ///
+    /// The value is guaranteed to be a value of `T` generated using [`FromType`] of the type that this
+    /// registration is used for.
     pub fn data<T: TypeData>(&self) -> Option<&T> {
         self.data
             .get(&TypeId::of::<T>())
             .and_then(|value| value.downcast_ref())
     }
 
-    /// Returns a mutable reference to the value of type `T` in this
-    /// registration's type data.
+    /// Inserts an instance of `D` into this registration's type data.
+    /// The value is created using the [`FromType<T>`] impl for `D`
+    /// and `T` must match the type that this [`TypeRegistration`] was created for.
```

Comment:
Maybe a `Panics` section to make it absolutely clear to users that this method can and will panic if the wrong type is used.